File: packages/react-native-nitro-modules/android/src/main/java/com/margelo/nitro/core/AnyMap.kt
package com.margelo.nitro.core

import androidx.annotation.Keep
import com.facebook.jni.HybridData
import com.facebook.proguard.annotations.DoNotStrip
import java.util.Dictionary

/**
 * Represents an untyped map of string keys with associated values.
 * This is like a JS [`object`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object).
 */
@Suppress("KotlinJniMissingFunction")
@Keep
@DoNotStrip
class AnyMap {
  private val mHybridData: HybridData

  /**
   * Create a new empty `AnyMap`.
   */
  constructor() {
    mHybridData = initHybrid()
  }

  /**
   * Create a new empty `AnyMap` with the given preallocated size
   */
  constructor(preallocatedSize: Int) {
    mHybridData = initHybrid(preallocatedSize)
  }

  /**
   * Create a new `AnyMap` from C++, which potentially already holds data.
   */
  @Suppress("unused")
  private constructor(hybridData: HybridData) {
    mHybridData = hybridData
  }

  companion object {
    /**
     * Converts the given [map] to a new [AnyMap].
     * @param map The map of keys/value types. Only a number of value types
     * are supported in [AnyMap] - see Nitro docs for more information.
     * @param ignoreIncompatible Whether incompatible key/value pairs should be ignored.
     * If this is `false`, an incompatible key/value pair will throw.
     */
    @JvmStatic
    external fun fromMap(
      map: Map<String, Any?>,
      ignoreIncompatible: Boolean,
    ): AnyMap
  }

  /**
   * Converts this [AnyMap] to a new [HashMap] by
   * copying each key/value.
   */
  external fun toHashMap(): HashMap<String, Any?>

  fun setAny(
    key: String,
    value: Any?,
  ) {
    setAnyValue(key, AnyValue.fromAny(value))
  }

  fun getAny(key: String): Any? {
    return getAnyValue(key).toAny()
  }

  private external fun fromHashMap(
    map: Map<String, Any?>,
    ignoreIncompatible: Boolean,
  )

  external fun contains(key: String): Boolean

  external fun remove(key: String)

  external fun clear()

  external fun getAllKeys(): Array<String>

  external fun isNull(key: String): Boolean

  external fun isDouble(key: String): Boolean

  external fun isBoolean(key: String): Boolean

  external fun isInt64(key: String): Boolean

  external fun isString(key: String): Boolean

  external fun isArray(key: String): Boolean

  external fun isObject(key: String): Boolean

  external fun getDouble(key: String): Double

  external fun getBoolean(key: String): Boolean

  external fun getInt64(key: String): Long

  external fun getString(key: String): String

  external fun getAnyArray(key: String): AnyArray

  external fun getAnyObject(key: String): AnyObject

  private external fun getAnyValue(key: String): AnyValue

  external fun setNull(key: String)

  external fun setDouble(
    key: String,
    value: Double,
  )

  external fun setBoolean(
    key: String,
    value: Boolean,
  )

  external fun setInt64(
    key: String,
    value: Long,
  )

  external fun setString(
    key: String,
    value: String,
  )

  external fun setAnyArray(
    key: String,
    value: AnyArray,
  )

  external fun setAnyObject(
    key: String,
    value: AnyObject,
  )

  private external fun setAnyValue(
    key: String,
    value: AnyValue,
  )

  external fun merge(other: AnyMap)

  private external fun initHybrid(): HybridData

  private external fun initHybrid(preallocatedSize: Int): HybridData
}

File: packages/nitro/cpp/core/AnyMap.cpp
//
// Created by Marc Rousavy on 30.07.24.
//

#include "AnyMap.hpp"

namespace margelo::nitro {

// Contains
bool AnyMap::contains(const std::string& key) const {
  return _map.contains(key);
}
void AnyMap::remove(const std::string& key) {
  _map.erase(key);
}
void AnyMap::clear() noexcept {
  _map.clear();
}

// Is
bool AnyMap::isNull(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    return false;
  }
  return std::holds_alternative<std::monostate>(found->second);
}
bool AnyMap::isDouble(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    return false;
  }
  return std::holds_alternative<double>(found->second);
}
bool AnyMap::isBoolean(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    return false;
  }
  return std::holds_alternative<bool>(found->second);
}
bool AnyMap::isBigInt(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    return false;
  }
  return std::holds_alternative<int64_t>(found->second);
}
bool AnyMap::isString(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    return false;
  }
  return std::holds_alternative<std::string>(found->second);
}
bool AnyMap::isArray(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    return false;
  }
  return std::holds_alternative<AnyArray>(found->second);
}
bool AnyMap::isObject(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    return false;
  }
  return std::holds_alternative<AnyObject>(found->second);
}

// Get
std::monostate AnyMap::getNull(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    throw std::runtime_error("The key \"" + key + "\" does not exist in this Map!");
  }
  if (auto result = std::get_if<std::monostate>(&found->second)) {
    return *result;
  } else {
    throw std::runtime_error("The value at key \"" + key + "\" is not a null!");
  }
}
double AnyMap::getDouble(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    throw std::runtime_error("The key \"" + key + "\" does not exist in this Map!");
  }
  if (auto result = std::get_if<double>(&found->second)) {
    return *result;
  } else {
    throw std::runtime_error("The value at key \"" + key + "\" is not a double!");
  }
}
bool AnyMap::getBoolean(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    throw std::runtime_error("The key \"" + key + "\" does not exist in this Map!");
  }
  if (auto result = std::get_if<bool>(&found->second)) {
    return *result;
  } else {
    throw std::runtime_error("The value at key \"" + key + "\" is not a bool!");
  }
}
int64_t AnyMap::getBigInt(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    throw std::runtime_error("The key \"" + key + "\" does not exist in this Map!");
  }
  if (auto result = std::get_if<int64_t>(&found->second)) {
    return *result;
  } else {
    throw std::runtime_error("The value at key \"" + key + "\" is not a bigint!");
  }
}
std::string AnyMap::getString(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    throw std::runtime_error("The key \"" + key + "\" does not exist in this Map!");
  }
  if (auto result = std::get_if<std::string>(&found->second)) {
    return *result;
  } else {
    throw std::runtime_error("The value at key \"" + key + "\" is not a string!");
  }
}
AnyArray AnyMap::getArray(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    throw std::runtime_error("The key \"" + key + "\" does not exist in this Map!");
  }
  if (auto result = std::get_if<AnyArray>(&found->second)) {
    return *result;
  } else {
    throw std::runtime_error("The value at key \"" + key + "\" is not an array!");
  }
}
AnyObject AnyMap::getObject(const std::string& key) const {
  auto found = _map.find(key);
  if (found == _map.end()) {
    throw std::runtime_error("The key \"" + key + "\" does not exist in this Map!");
  }
  if (auto result = std::get_if<AnyObject>(&found->second)) {
    return *result;
  } else {
    throw std::runtime_error("The value at key \"" + key + "\" is not an object!");
  }
}

// Set
void AnyMap::setNull(const std::string& key) {
  _map.emplace(key, std::monostate());
}
void AnyMap::setDouble(const std::string& key, double value) {
  _map.emplace(key, value);
}
void AnyMap::setBoolean(const std::string& key, bool value) {
  _map.emplace(key, value);
}
void AnyMap::setBigInt(const std::string& key, int64_t value) {
  _map.emplace(key, value);
}
void AnyMap::setString(const std::string& key, const std::string& value) {
  _map.emplace(key, value);
}
void AnyMap::setArray(const std::string& key, const AnyArray& value) {
  _map.emplace(key, value);
}
void AnyMap::setObject(const std::string& key, const AnyObject& value) {
  _map.emplace(key, value);
}
void AnyMap::setAny(const std::string& key, const AnyValue& value) {
  _map.emplace(key, value);
}

// C++ getter
const std::unordered_map<std::string, AnyValue>& AnyMap::getMap() const {
  return _map;
}

} // namespace margelo::nitro
